File: HW 2 (Java)/src/Rectangle.java
import java.util.List;
import java.util.Objects;

public class Rectangle extends Quadrilateral implements SymmetricTwoDShape {

    public Rectangle(double... vertices) {
        this(TwoDPoint.ofDoubles(vertices));
    }


    public Rectangle(List<TwoDPoint> vertices) {
        super(vertices);
    }


    /**
     * The center of a rectangle is calculated to be the point of intersection of its diagonals.
     *
     * @return the center of this rectangle.
     */
    @Override
    public Point center() {
        double side1 = this.getSideLengths()[0]/2.0;
        double side2 = this.getSideLengths()[1]/2.0;
        double x = this.getPosition().get(0).getX() - side1;
        double xRound = (Math.round(x * 1000.0) / 1000.0);
        double y = this.getPosition().get(0).getY() - side2;
        double yRound = (Math.round(y * 1000.0) / 1000.0);

        TwoDPoint rectangleCenter = new TwoDPoint(xRound, yRound);
        return rectangleCenter; // TODO
    }

    @Override
    public boolean isMember(List<? extends Point> vertices) {

        if (! super.isMember(vertices))
            return false;

        for (int i = 0; i< vertices.size(); i++){
            if (! (vertices.get(i) instanceof TwoDPoint))
                return false;
        }
        if ( vertices.get(0).equals(vertices.get(1)) || vertices.get(0).equals(vertices.get(2))
                || vertices.get(0).equals(vertices.get(3)) || vertices.get(1).equals(vertices.get(2))
        || vertices.get(1).equals(vertices.get(3)) || vertices.get(2).equals(vertices.get(3))
        )
            return false;

        Double side1 = this.getSideLengths()[0];
        Double side1Round = (Math.round(side1 * 1000.0) / 1000.0);
        Double side2 = this.getSideLengths()[1];
        Double side2Round = (Math.round(side2 * 1000.0) / 1000.0);
        Double side3 = this.getSideLengths()[2];
        Double side3Round = (Math.round(side3 * 1000.0) / 1000.0);
        Double side4 = this.getSideLengths()[3];
        Double side4Round = (Math.round(side4 * 1000.0) / 1000.0);

        Double diagonol1 = Math.pow(side1 * side1 + side4 * side4 , 0.5);
        Double diagonol1Round = (Math.round(diagonol1 * 1000.0) / 1000.0);
        Double diagonol2 = Math.pow(side2 * side2 + side3 * side3 , 0.5);
        Double diagonol2Round = (Math.round(diagonol2 * 1000.0) / 1000.0);


        if (side1Round.equals(side3Round) && side2Round.equals(side4Round)) {
            if (diagonol1Round.equals(diagonol2Round));
            return true;
        }

        return false;
    }

    @Override
    public double area() {
        return super.area(); // TODO
    }
}

File: HW 3 (Python)/hw3-codebase/rectangle.py
from quadrilateral import Quadrilateral
from two_d_point import TwoDPoint


class Rectangle(Quadrilateral):

    def __init__(self, *floats):
        super().__init__(*floats)
        if not self.__is_member():
            raise TypeError("A {} cannot be formed by the given coordinates.".format(self.__class__.__name__))

    def __is_member(self):
        """Returns True if the given coordinates form a valid rectangle, and False otherwise."""
        if ((self.vertices[0] == self.vertices[1]) or (self.vertices[0] == self.vertices[2]) or (self.vertices[0] == self.vertices[3]) or (self.vertices[1] == self.vertices[2]) or (self.vertices[1] == self.vertices[3]) or (self.vertices[2] == self.vertices[3])):
            return False
        side1 = round(self.side_lengths()[0],5)
        side2 = round(self.side_lengths()[1],5)
        side3 = round(self.side_lengths()[2],5)
        side4 = round(self.side_lengths()[3],5)
        if ((side1 == side3) and (side2 == side4)):
            if ( (round((pow(side1,2) + pow(side2,2)),5)) == ( round((pow(side3,2) + pow(side4,2)),5))):
                return True
        return False


    def center(self):
        newX = self.vertices[0].x + self.vertices[1].x + self.vertices[2].x + self.vertices[3].x
        newY = self.vertices[0].y + self.vertices[1].y + self.vertices[2].y + self.vertices[3].y

        return TwoDPoint(newX/4.0, newY/4.0)  # TODO

    def area(self):
        return self.side_lengths()[0] * self.side_lengths()[1]  # TODO

    def __str__(self) -> str:
        return 'I am a Rectangle with ' + str(self.vertices[0]) + " , "+ str(self.vertices[1]) + " , " + str(self.vertices[2]) + " , " + str(self.vertices[3])


    def __eq__(self, other: object) -> bool:
        if isinstance(other, Rectangle):
            if ((self.vertices[0] == other.vertices[0]) and (self.vertices[1] == other.vertices[1]) and (
                    self.vertices[2] == other.vertices[2]) and (self.vertices[3] == other.vertices[3])):
                return True
        return False  # TODO


    def __ne__(self, other: object) -> bool:
        return not self.__eq__(other)
